[kylieluk88] iP

kylieluk88 opened the pull request:
### Hailey

> “Time is the scarcest resource and unless it is managed nothing else can be managed.” – Peter Ducker 

Hailey helps you keep track of tasks so you can better managed your time. It's,

- text-based 
- easy to learn
- FAST to use

All you need to do is,
download it from [here](https://github.com/kylieluk88/ip).

1. double-click it. 
2. add your tasks.
3. let it manage your tasks for you!

And it is **FREE**!

Features:

-  Managing tasks ☑️☑️
-  Managing deadlines 📅📅
-  Reminders (coming soon) 🔔🔔

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public static void main(String[] args) throws IOException { 
        new Hailey("data/tasks.txt").run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

kylieluk committed:
Add greeting from Hailey

kylieluk committed:
Add support for echo

kylieluk committed:
Add UI class and addTask function

kylieluk committed:
Add TaskList class to add and print tasks

kylieluk committed:
Add support for task list functions

kylieluk committed:
Add Task class

kylieluk committed:
Add support for Mark and Unmark as done

kylieluk committed:
Add Deadline, Event, Todo subclasses, update main to support

kylieluk committed:
Update UI to support different tasks

kylieluk committed:
Add exception handling for empty description fields and unknown input

kylieluk committed:
Add support for deleting tasks from the list, changed TaskList from an array to ArrayList

kylieluk committed:
Debug error message

kylieluk committed:
Add A-TextUiTesting for todo, deadline, list, mark, and error catching

kylieluk committed:
Delete old files

kylieluk committed:
Add functions to read and write file

kylieluk committed:
Debug file reading function

kylieluk committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Debug file reading function
  Add functions to read and write file

kylieluk committed:
Modify start and end time in Deadline and Event classes

kylieluk committed:
Debug date and time formatting in main

kylieluk committed:
Revert "Debug date and time formatting in main"

This reverts commit 87c7eb09a73ad70102881416fb1abc4e33b980bb.

kylieluk committed:
Add support for date and time

kylieluk committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Revert "Debug date and time formatting in main"
  Debug date and time formatting in main
  Modify start and end time in Deadline and Event classes

kylieluk committed:
Add Storage class

kylieluk committed:
Add Parser class and edit Hailey to support

kylieluk committed:
Debug Storage functions

kylieluk committed:
Move files to respective packages

kylieluk committed:
Merge commit '81a9c537afe8492675eb253968939364d65db091'

* commit '81a9c537afe8492675eb253968939364d65db091':
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

kylieluk committed:
Add J-Unit tests for Parser and TaskList

kylieluk committed:
Add JavaDoc comments

kylieluk committed:
Revert "Add J-Unit tests for Parser and TaskList"

This reverts commit b68260a4d4ad5b1f19043fede2908dfdc871d9d3.

kylieluk committed:
Add JavaDoc comments

kylieluk committed:
Rename UI.java to Ui.java to comply with coding standards

kylieluk committed:
Rename UI to comply with coding standard

kylieluk committed:
Remove files during packaging

kylieluk committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments

kylieluk committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Rename UI to comply with coding standard
  Rename UI.java to Ui.java to comply with coding standards

kylieluk committed:
Add support for finding tasks

kylieluk committed:
Add support for Find task

AnAfterthought commented on src/main/java/hailey/ui/Ui.java:
```diff
@@ -0,0 +1,69 @@
+package hailey.ui;
+
+import hailey.task.*;
```
Should list imports explicitly.

AnAfterthought commented on src/main/java/hailey/ui/Ui.java:
```diff
@@ -0,0 +1,69 @@
+                + "What can I do for you?\n" + line);
+    }
+
+    public void help() {
```
Perhaps rename as ```showHelp()```?

AnAfterthought commented on src/main/java/hailey/task/ToDo.java:
```diff
@@ -0,0 +1,31 @@
+     * Returns a formatted string for saving the Event task to a file.
+     * @return The formatted save string.
+     */
+    public String saveFormat() {
```
```saveFormat()``` feels like a noun. Perhaps ```toSaveFormat()``` will be clearer, similar to ```toString()```.

AnAfterthought commented on src/main/java/hailey/task/TaskList.java:
```diff
@@ -0,0 +1,120 @@
+     * Gets the total number of tasks in the list.
+     * @return The count of tasks.
+     */
+    public int getCount() {
```
Perhaps ```getSize()``` is better, since you are calling ```tasks.size()```.

AnAfterthought commented on src/main/java/hailey/ui/Ui.java:
```diff
@@ -0,0 +1,69 @@
+        System.out.println(line + "Bye. Hope to see you again soon!\n" + line);
+    }
+
+    public void addTask(Task task, TaskList tasks) {
```
It appears that this method is simply printing a message. Perhaps it would be better to call it ```printAddMessage()```?

AnAfterthought commented on src/main/java/hailey/ui/Ui.java:
```diff
@@ -0,0 +1,69 @@
+        System.out.println(line + "Here are the tasks in your list:\n" + tasks.printTasks() + line);
+    }
+
+    public void markDone() {
```
It appears that this method is simply printing a message. Perhaps it would be better to call it printMarkMessage()?

AnAfterthought commented on src/main/java/hailey/ui/Ui.java:
```diff
@@ -0,0 +1,69 @@
+                "\nNow you have " + (tasks.getCount() - 1) + " tasks in the list.\n" + line);
+    }
+
+    public void errorMessage() {
```
```errorMessage()``` is a noun. Perhaps change it to ```showErrorMessage()```?

AnAfterthought commented on src/main/java/hailey/parser/Parser.java:
```diff
@@ -0,0 +1,102 @@
+     * @throws HaileyException If an invalid command is encountered.
+     */
+    public boolean processCommand(String input, TaskList tasks, Ui ui, Storage storage) throws HaileyException {
+        if (input.equals("bye")) {
```
Perhaps consider using switch statement?

AnAfterthought reviewed (commented):
There does not appear to be too many coding violations so good job! 👍 
However, looking at your UI class, it seems to only be printing messages. Perhaps you should use method names to reflect that? Some of the current names make it seem as though the UI class has some logic to manage tasks.

Emmanuel2001 commented on src/main/java/hailey/Hailey.java:
```diff
@@ -0,0 +1,63 @@
+import java.io.*;
+import java.util.Scanner;
+
+/**
+ * The main class of the Hailey chatbot application.
+ * Initializes components and runs the chatbot.
+ */
```
I like how you have header comments for each public class!

Emmanuel2001 commented on src/main/java/hailey/Hailey.java:
```diff
@@ -0,0 +1,63 @@
+        tasks = storage.readFile();
+        ui.greet();
+        Scanner scanner = new Scanner(System.in);
+        boolean isRunning = true;
```
I like how the boolean is named like booleans!

Emmanuel2001 commented on src/main/java/hailey/parser/Parser.java:
```diff
@@ -0,0 +1,102 @@
+            if (!input.contains("/by")) {
+                throw new HaileyException("oops! deadline must include '/by' keyword");
+            }
+            String[] parts = input.substring(9).split(" /by ");
+            if (parts.length < 2) {
+                throw new EmptyDescriptionException("deadline");
+            }
+            LocalDateTime by = parseDateTime(parts[1]);
+            Deadline deadline = new Deadline(parts[0], by);
+            tasks.addTask(new Deadline(parts[0], by));
+            ui.addTask(deadline, tasks);
+        } else if (input.startsWith("event ")) {
+            if (!input.contains("/from") || !input.contains("/to")) {
+                throw new HaileyException("oops! event must include '/from' and '/to' keyword");
+            }
+            String[] parts = input.substring(6).split(" /from | /to ");
+            if (parts.length < 3) {
+                throw new EmptyDescriptionException("event");
+            }
+            LocalDateTime start = parseDateTime(parts[1]);
+            LocalDateTime end = parseDateTime(parts[2]);
+            Event event = new Event(parts[0], start, end);
+            tasks.addTask(event);
+            ui.addTask(event, tasks);
+        } else if (input.startsWith("todo")) {
+            if (input.substring(4).isEmpty()) {
+                throw new EmptyDescriptionException("todo");
+            }
+            String description = input.substring(5);
+            ToDo todo = new ToDo(description);
+            tasks.addTask(todo);
+            ui.addTask(todo, tasks);
+        } else if (input.startsWith("find ")) {
+            String keyword = input.substring(5).trim();
+            ArrayList<Task> matchingTasks = tasks.find(keyword);
+            ui.showMatchingTasks(matchingTasks);
+        } else {
+            throw new HaileyException("Sorry, didn't quite catch that. What did you say?");
+        }
+        return true;
```
The code block for the if else statement looks neat with correct indentation, nice work! 
Perhaps, you could try using switch case?

Emmanuel2001 commented on src/main/java/hailey/storage/Storage.java:
```diff
@@ -0,0 +1,108 @@
+    /**
+     * Ensures that the storage file exists. Creates it if necessary.
+     */
+    private void checkFileExists() {
```
I like how the methods' names follow the naming convention and describes what it does. Good job!

Emmanuel2001 reviewed (commented):
Hi Kylie, you did a good job overall! I can tell that you have put in effort into ensuring that your code is of good quality and standard :)

codebreaker64 commented on src/main/java/UI.java:
Good job on following the coding Standard! It would be even better if you can put the file in a package called UI under your hailey package and include java header for all of them.

codebreaker64 commented on src/main/java/UI.java:
```diff
@@ -0,0 +1,56 @@
+import org.w3c.dom.ls.LSOutput;
+
+public class UI {
```
I think the class suppose to be private, so cannot put public.

codebreaker64 commented on src/main/java/hailey/Hailey.java:
```diff
@@ -0,0 +1,63 @@
+ * The main class of the Hailey chatbot application.
+ * Initializes components and runs the chatbot.
+ */
+public class Hailey {
```
Should omit the public visibility.

codebreaker64 commented on src/main/java/hailey/exception/EmptyDescriptionException.java:
```diff
@@ -0,0 +1,7 @@
+package hailey.exception;
+
+public class EmptyDescriptionException extends HaileyException {
```
Should omit the public visibility

codebreaker64 commented on src/main/java/hailey/exception/HaileyException.java:
```diff
@@ -0,0 +1,8 @@
+package hailey.exception;
+
+public class HaileyException extends Exception {
```
Should omit public visibility

codebreaker64 commented on src/main/java/hailey/parser/Parser.java:
```diff
@@ -0,0 +1,102 @@
+/**
+ * The Parser class interprets user input and executes the corresponding commands.
+ */
+public class Parser {
```
Should omit public visibility.

codebreaker64 commented on src/main/java/hailey/storage/Storage.java:
```diff
@@ -0,0 +1,108 @@
+/**
+ * The Storage class handles loading and saving tasks from a file.
+ */
+public class Storage {
```
Should omit public visibility.

codebreaker64 commented on src/main/java/hailey/task/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Represents a task with a specific deadline.
+ */
+public class Deadline extends Task {
```
should omit public visibility.

codebreaker64 commented on src/main/java/hailey/task/Event.java:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Represents a task that occurs within a specific time range.
+ */
+public class Event extends Task {
```
Should omit public visibility

codebreaker64 commented on src/main/java/hailey/task/Task.java:
```diff
@@ -0,0 +1,58 @@
+package hailey.task;
+import hailey.exception.HaileyException;
+
+public class Task {
```
Should omit public visibility

codebreaker64 commented on src/main/java/hailey/task/TaskList.java:
```diff
@@ -0,0 +1,120 @@
+/**
+ * Manages a list of tasks.
+ */
+public class TaskList {
```
Should omit public visibility

codebreaker64 commented on src/main/java/hailey/task/ToDo.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Represents a simple task without a deadline or time range.
+ */
+public class ToDo extends Task {
```
Should omit public visibility

codebreaker64 commented on src/main/java/hailey/ui/Ui.java:
```diff
@@ -0,0 +1,69 @@
+
+import java.util.ArrayList;
+
+public class Ui {
```
Should omit public visibility

codebreaker64 reviewed (commented):
Good job on closely following the coding standard. A few things I would change is the UI.java file should be in a package called UI under hailey, include java headers for all the methods and omit all the public visibility(Only data class with no behaviour can set to public).

kylieluk committed:
Add support for GUI

kylieluk committed:
Debug error messages

kylieluk committed:
Add assertions for TaskList

kylieluk committed:
Refactor processCommand to use switch statement

processCommand previously used an if-else chain, making it less readable
and harder to maintain as new commands were added.

Using a switch statement improves:
* Readability by reducing deeply nested conditions.
* Maintainability by making it easier to add new commands.
* Performance by enabling optimized branching.

Let's refactor processCommand to use a switch statement while preserving
all existing functionality. Command parsing logic remains unchanged, and
validation checks for task indices are still in place.

kylieluk88 committed:
Merge pull request #1 from kylieluk88/branch-A-Assertions

Add assertions for TaskList

kylieluk committed:
Modify code to comply with coding standards

Some code that was needed in previous iterations have become
 redundant due to new modifications. This made the code longer
than necessary. Some code also violated coding standards.

Let's clean up the code and make sure it follows coding standards
for clarity of reading.

kylieluk committed:
Merge commit '807092caecd273556c41266030a849d7acc3e2c3' into branch-A-CodeQuality

* commit '807092caecd273556c41266030a849d7acc3e2c3':

kylieluk committed:
Debug file storage to ensure tasks.txt is saved correctly

After addition of GUI file was not being written into after each command.

Let's update the code to ensure tasks are being saved in the txt file
with each user input.

kylieluk committed:
Add support to detect duplicate tasks with same description

kylieluk committed:
Merge branch 'branch-C-DetectDuplicates'

* branch-C-DetectDuplicates:
  Add support to detect duplicate tasks with same description

kylieluk committed:
Fix package name to match naming convention

kylieluk committed:
Modify processCommand() to shorten method

kylieluk committed:
Delete redundant code

kylieluk committed:
Edit responses to include personality

kylieluk committed:
Add a User Guide

kylieluk committed:
Edit User Guide to include product screenshot

kylieluk committed:
Add product screenshot to docs folder

kylieluk committed:
Edit User Guide to correctly show product screenshot

kylieluk committed:
Update build.gradle to match file names

kylieluk committed:
Debug to solve JavaFX runtime error

kylieluk committed:
Edit style of chatbot responses
